### ISSUE TITLE
Poor performance on custom nodes with many ports

### ISSUE THREAD

**Issue by ghost** [Blazor-Diagrams/Blazor.Diagrams#110]
During evaluation we noticed that a custom node containing a rather large amount of ports (76 during my testing) can result in very poor performance while adding that node type to the diagram canvas.

We narrowed the cause down to the PortRenderer implementation doing JSInterop during its initial render and another one following that caused by the JS OnResize event resulting in a whopping 152 JSI invocations for a single node. The same behaviour of the two initial rendering cycles can be observed on the NodeRenderer itself causing additional rendering overhead.

Even though we found a solution to that problem there would be rather dramatic changes ([see here](https://github.com/xevoryn/Blazor.Diagrams/tree/perf-opt/port-renderer)) to the Diagrams library itself. Is there any other way to increase performance during rendering? And is the [approach](https://github.com/xevoryn/Blazor.Diagrams/tree/perf-opt/port-renderer) acceptable?

EDIT: I updated the links because I transferred the fork to another account that's more usable for me. I am continuing to write and contribute with the account @xevoryn.

**Comment by zHaytam** [comment]
Hello,

This was my concern for quite some time now. The number of JS calls gets higher with every new node and port, so it's only fair that the performance drops in your case.

The solution that I was thinking of is basically the same as yours, except more general. I wanted to batch all JS calls into one (or more over a period of time), which should fix the performance issue you're seeing.

Other than this problem, I don't think there is another bottleneck elsewhere. Maybe during the mouse move event, but that's still untested so I don't really know how it will be.

Can you put this test in a public repo if possible? I would like to see how slow is it and how your fix helps. Maybe add a test page to your fork/branch and I can see everything there?

Thank you for taking the time to check the library and the problem.

**Comment by xevoryn** [comment]
So here we go...

I've created a new [branch](https://github.com/xevoryn/Blazor.Diagrams/tree/perf-opt/unoptimized-testpage) that contains our component and added it to the demo pages. It can be found under "Large Node Performance Demo". It contains an input field to specify the number of nodes to generate and three buttons:

1. Create a small block (for reference)
2. Create the large block I mentioned before with 76 ports
3. Create the amount of both blocks

You will surely see that the performance suffers while generating 50 of the large blocks which was our test case. Please ignore the fact that I updated your sample projects to dotnet 5. I had to do this because we wrote all of our demo CSS scoped and I was faster updating the projects rather then re-writing all of the CSS. Tomorrow I will provide you another branch which implements our custom node and the PortRenderer changes I made in the branch mentioned before. I let you know when it's finished.

**Comment by xevoryn** [comment]
Ok, ich implemented the test page into the [PortRenderer branch](https://github.com/xevoryn/Blazor.Diagrams/tree/perf-opt/port-renderer) mentioned before. So there are now two branches one which does not have the improvements and the other one has. 

What is wondering me is that the improvements done to the PortRenderer are not that significant in your demo page as in our one. We achieved an improvement of ~30%. But I guess the other improvements we did to our node component, our more simple test page and some other improvements I did to the lib are helping as well.

If the branches do not help you let me know so I can provide another example.

**Comment by xevoryn** [comment]
I extracted our demo solution from another repo an made it public [right here](https://github.com/xevoryn/Blazor.Diagrams.Demo). Just start the `DiagramDemo.Server` and connect from your browser to `http://localhost:57000`.

If you checkout commit after commit (starting [here](https://github.com/xevoryn/Blazor.Diagrams.Demo/commit/a25219354df06aa6f4a19fa2a4c230956417af82)) you will see that the performance of the 50 large (Sonos) Blocks increases with each one of it. It leads to the point where I am (utilizing a dual core CPU) down from ~14sec of processing time to ~2sec. Let me know what you think about the approaches we took to increase library performance.

**Comment by zHaytam** [comment]
I tried both your sample projects and I'm very confused right now...
In the fork project, both branches have the same performance (at least visually, I didn't time anything).
In your other project, I can only see the nodes after ~6s, which is the same as the fork.

Am I doing something wrong? (This is on a Ryzen 7 3700x)

**Comment by xevoryn** [comment]
Very interesting... So I have two questions:

- What browser are you using?
- Did you have the dev console open while testing?

We also observed a already known bug in chromium where opening the console does slows down performance when using observers dramatically. And that's also the case for Blazor.Diagrams. So you have to leave the console closed while testing.

**Comment by zHaytam** [comment]
- Google Chrome
- Nope

I know, with the dev console opened, the loading takes ~33 seconds lol. It's insane. Once I finish some urgent irl stuff, I will investigate this in details.

**Comment by xevoryn** [comment]
So may I update that Issue once again: I tried my demo solution on three different machines:

- M1: [intel core i5 7260U](https://ark.intel.com/content/www/de/de/ark/products/97539/intel-core-i5-7260u-processor-4m-cache-up-to-3-40-ghz.html) up to 3.9GHz
- M2: [intel core i5 2410M](https://ark.intel.com/content/www/us/en/ark/products/52224/intel-core-i5-2410m-processor-3m-cache-up-to-2-90-ghz.html) up to 2.9GHz
- M3: [intel core i7 7700K](https://ark.intel.com/content/www/us/en/ark/products/97129/intel-core-i7-7700k-processor-8m-cache-up-to-4-50-ghz.html) up to 4.5GHz

On every one I got a significant increase in performance over the commits of the master-branch. I would be very interested in why it's running that weird on your machine. Let me know when you had the time to analyse it.

**Comment by zHaytam** [comment]
Hi,

I will start investigating this again. Sorry for the big delay!

**Comment by zHaytam** [comment]
Update:
Batching the JS call causes crashes in Server Side Blazor. This is due to the limit of how much data can the server receive from JS (32Kb by default). For 1500+ ports, we get to ~200Kb, which is quite a lot for 1 call...

Maybe a better solution would be to have a new type of ports, static ones that have a fixed size, so no need to do JS INterop to retrieve the size. This sounds like a good solution since most ports are fixed/static.

**Comment by xevoryn** [comment]
Yes, I have heard of that limitation but I haven't used Blazor Server before and I didn't tested that. Sorry for that.

Static ports sounds like a great solution to the problem. Maybe they should be the default type of port if there are probably more static than dynamic ports as you've said.

**Comment by zHaytam** [comment]
@xevoryn After another look at your ports, I'm not sure static ports would work in your case. What do you think?

By static ports I mean something like:
```cs
var node = new MyCustomNode();
node.AddPort(new StaticPort(Position.Relative, new Position(10, 20)));
// ^ This would position the port at (Node.X + 10, Node.Y + 20)
```

As you can see, you will need to know the position (absolute or relative to the node) beforehand. Do you think you can do that? Please share your thoughts with me.

As for JS batching, I will be adding an option that you can enable/disable. I am still hoping to get this static ports feature done in order to remove the need for JS all together, which should increase the performance by ~30% or plus.

**Comment by xevoryn** [comment]
We are aligning the ports without giving them a specific position and we are not changing them during runtime. So if the position parameter is optional like the one on the port right now it would perfectly fine for us. If it isn't possible to make it optional a position of (0, 0) would do as well because we are not using it anyway in our custom node implementation.

**Comment by zHaytam** [comment]
That would mean the port still needs to fetch the position after the CSS is applied, because that's where the links go from/to.

**Comment by xevoryn** [comment]
Ok, I see where you are going but setting a position should not be a problem for us.

**Comment by xhydongda** [comment]
Accoding to [MS document](https://docs.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-6.0),  **a rendering overhead of around 0.06 ms per component instance was seen in a Blazor WebAssembly app**,  reducing components would be the right way.

**Comment by zHaytam** [comment]
Hello, working on 3.0 currently and I'm checking this issue again.
You said it's fine for you to set a position, but how would you know it without css? I see you're using a grid for placements so what will you do if static ports were a thing?
